testsuite/gateway/gateway_api/gateway.py

```diff
@@ -1,59 +1,61 @@
 """Module containing all gateway classes"""
 
-from typing import Any
+from typing import Any, Optional
 
 import openshift_client as oc
 
 from testsuite.certificates import Certificate
-from testsuite.gateway import Gateway
+from testsuite.gateway import Gateway, GatewayListener
 from testsuite.kubernetes.client import KubernetesClient
-from testsuite.kubernetes import KubernetesObject
+from testsuite.kubernetes import KubernetesObject, modify
 from testsuite.kuadrant.policy import Policy
-from testsuite.utils import check_condition
+from testsuite.utils import check_condition, asdict
 
 
 class KuadrantGateway(KubernetesObject, Gateway):
     """Gateway object for Kuadrant"""
 
+    # Used to make unique listener names
+    listener_counter = 1
+
     @classmethod
-    def create_instance(cls, cluster: KubernetesClient, name, hostname, labels, tls=False):
+    def create_instance(cls, cluster: KubernetesClient, name, listener: GatewayListener, labels):
         """Creates new instance of Gateway"""
 
         model: dict[Any, Any] = {
             "apiVersion": "gateway.networking.k8s.io/v1beta1",
             "kind": "Gateway",
             "metadata": {"name": name, "labels": labels},
-            "spec": {
-                "gatewayClassName": "istio",
-                "listeners": [
-                    {
-                        "name": "api",
-                        "port": 80,
-                        "protocol": "HTTP",
-                        "hostname": hostname,
-                        "allowedRoutes": {"namespaces": {"from": "All"}},
-                    }
-                ],
-            },
+            "spec": {"gatewayClassName": "istio", "listeners": [asdict(listener)]},
         }
 
-        if tls:
-            model["spec"]["listeners"] = [
-                {
-                    "name": "api",
-                    "port": 443,
-                    "protocol": "HTTPS",
-                    "hostname": hostname,
-                    "allowedRoutes": {"namespaces": {"from": "All"}},
-                    "tls": {
-                        "mode": "Terminate",
-                        "certificateRefs": [{"name": f"{name}-tls", "kind": "Secret"}],
-                    },
-                }
-            ]
-
         return cls(model, context=cluster.context)
 
+    @modify
+    def add_listener(self, listener: GatewayListener):
+        """Adds a listener to Gateway. If the argument doesn't contain name, generate new sequential one."""
+        if listener.name is None:
+            listener.name = f"api{self.listener_counter}"
+            self.listener_counter += 1
+        self.model.spec.listeners.append(asdict(listener))
+
+    @modify
+    def remove_listener(self, hostname: GatewayListener | str):
+        """Removes a listener based on hostname"""
+        if isinstance(hostname, GatewayListener):
```

Comment:
Do we need the option to remove the listener based on `GatewayListener`? I don't like `hostname: GatewayListener`

Reply:
I changed the logic, now listener name is used to find listener to remove